### PR TITLE
Create new lane types to express different types of buffers for prote…

### PR DESCRIPTION
**Pull request by dabreegster**
…cted bike lanes. They're only created manually right now, to explore rendering.

The goal is to start modeling different types of bike lanes -- some are right in the door-zone of street parking, others have flimsy little flex posts protecting from cars, but some have proper barriers in place. Eventually we'll use this in the routing and simulation layer. But the goal for this PR is actually just to explore the rendering style.

I've started with 4 types of barriers, but happy to add more!

For each of these, any better ideas for visualizing in our top-down view?

# Stripes

Top-down rendering:
Screenshot from 2021-07-16 14-50-25
Reference:
buffered-bike-lane_street_0

# Flex posts ("wands" in the UK)

Rendering:
Screenshot from 2021-07-16 14-51-45
Reality:
bikes_lanes_slider

# Planters

Rendering:
Screenshot from 2021-07-16 14-52-52
Reality:
2016_0211_2ndAvePioneerSquareFlyer-concept

# Jersey barrier

Rendering:
Screenshot from 2021-07-16 14-54-00
Reality:
bikelane-protected-jersey-barrier-flushing-ave--fortheloveofbikes-thumb

# Next steps

There's more work needed to create/modify these properly in the road editor UI. I also want to start tagging these in OSM, and then importing into A/B Street. https://wiki.openstreetmap.org/wiki/Berlin/Verkehrswende/Radwege#Tagging-Beispiele seems to have some traction.

### CODE REVIEW

**Comment by dabreegster** [comment]
@trangrei, @matkoniecz, @westnordost -- any design ideas for drawing these bike buffers in a ~top-down view? Or other types of buffers that should be included with this first draft?

**Comment by westnordost** [comment]
You seem to have left out the most important segregation: A curb, so the cycle lane becomes a cycle track.

The drafts look fine to me, I'd maybe make the planters more bush-like for better visual distinction (after all the planters don't need to be rectangular)
image

**Comment by michaelkirk** [comment]
>  I'd maybe make the planters more bush-like for better visual distinction 

Those plants look nice, though I don't think we have a great way for drawing sprites like that. We do have textures at our disposal, but they may be cropped arbitrarily.

**Comment by dabreegster** [comment]
> You seem to have left out the most important segregation: A curb, so the cycle lane becomes a cycle track.

Ah, true! Any thoughts how to create some sense of depth?
Screenshot from 2021-07-16 14-54-00

> though I don't think we have a great way for drawing sprites like that.

Those plants do look nice! And we can draw them without too much trouble -- create an SVG of the plant, then place it like a stencil every few meters, the same as we do with [bike stencils](https://github.com/a-b-street/abstreet/blob/d142b4fbd076e4bdcb8a5a73cba4c1e5e311a408/map_gui/src/render/lane.rs#L92)

**Comment by dabreegster** [comment]
I found some old concept art that I had commissioned in the early days:
Screenshot from 2021-07-16 16-59-44
Screenshot from 2021-07-16 17-00-23

**Comment by westnordost** [comment]
Or use a continuous white line before the gutter to mark the end of a street and the beginning of a sidewalk

image

**Comment by matkoniecz** [comment]
Jersey barrier looks flat in this example right now, making it more 3D would be nice.

> Planters

I would change green shade, right now it has shade of toxic waste from cartoon. Some symbolic vegetation as mentioned also would be nice.

> Flex posts ("wands" in the UK)

I would make dost smaller and denser, right now it looks like a row of advertising columns

> Stripes

No idea is there OSM tagging for that but sometimes entire lanes are striped out and not drivable.

> curb

Some lanes have both curb and bollard separation

**Comment by Robinlovelace** [comment]
Linking to #712 it seems planters can be linear features. I agree with comments in https://github.com/a-b-street/abstreet/pull/704#issuecomment-881797422 that a leafier shade of green could be good but just getting them in there will be amazing and, by the looks of it, that's what this PR does. Amazing work.

**Comment by Robinlovelace** [comment]
This PR https://github.com/a-b-street/abstreet/pull/713 makes the planter look like this:

image

**Comment by Robinlovelace** [comment]
Inspiration: https://wiki.openstreetmap.org/wiki/Tag:barrier%3Dplanter

**Comment by Robinlovelace** [comment]
Looking forward to seeing this rendered properly in a digital map (they are ignored completely in all OSM renderers I've seen): 

image

There is a video of them somewhere, a work of art that is as of yesterday [represented on OSM ](https://www.openstreetmap.org/way/965986113)as follows (invisible unless you select the way):

image

**Comment by dabreegster** [comment]
I think the first cut of this is done. There are 5 types of buffers now available in the road editor:
Screenshot from 2021-07-21 13-05-32
(Yes, those icons need some work.)

Stripes:
Screenshot from 2021-07-21 13-06-28

Flex posts:
Screenshot from 2021-07-21 13-06-54

Planters (need to center between the stripes better):
Screenshot from 2021-07-21 13-07-22

Jersey barrier:
Screenshot from 2021-07-21 13-07-55

And just a raised curb:
Screenshot from 2021-07-21 13-08-45

Some of my next steps:
- improve rendering and the editor icons (with help)
- rethink the road editor UI. Treating all of these as regular lane types is weird. Maybe there should be a single barrier dropdown, or a way to denote a category of lane. Or maybe the barrier lanes should be "options" on a bike lane/sidewalk.
- Investigate OSM tagging and start picking these up during map imports

**Comment by dabreegster** [comment]
https://twitter.com/redblobgames/status/1418081197037473798 raised a great idea -- just cheat and stick in a quasi-side view of an object, to create a better sense of depth. (We do this for stop signs already.) All the curvy roads at every angle could get tricky, but the simple cases work well:
Screenshot from 2021-07-21 13-06-28
Screenshot from 2021-07-16 14-50-25
Screenshot from 2021-07-22 08-16-15

**Review by michaelkirk — APPROVED** [review]
The code looks good - though some of the actual assets could be cleaned up.

- the diagonal striping should go all the way into their containing rect
- the planter's look like their corners are chopped off

With this iteration of the jersey barriers, the color does help with the depth perception a bit, but they're still a little hard to distinguish. the "quasi-side view" approach might help. Or drawing a thin line to delineate segments, or slightly rotating each segment so that they don't look like a 100% contiguous line... 

I wouldn't consider any of this blocking feedback though. Maybe they'd make good "starter" / "help wanted" tasks?

**Comment by michaelkirk** [comment]
> just cheat and stick in a quasi-side view of an object

I think this could work for more things. I think in your particular example, I don't like the look of it, but that has more to do with the particular photo-realistic asset you choses which looks very out of place with the rest of the ABStreet style.

**Comment by Robinlovelace** [comment]
Just catching up with this now. Overall I think the main point is that this PR enables people to add separation between modes. That is hugely important, and a key element of streetmix. My take is that the fact that the basic functionality exists is more important than the aesthetics, and that seems to be implemented based on this screenshot which looks amazing:

image

Only question: do all these great new options clutter the UI? Less can be more.

**Comment by westnordost** [comment]
Or for cycleways behind kerbs, like this:


<b>Cycle track</b> (=cycleway separated by kerb): The paving on the cycle track should make it clear that the track is off the roadway. There is a slight 3D effect for the kerb

File: game/src/edit/roads.rs [hunk 1]
```diff
@@ -464,21 +464,26 @@ fn make_main_panel(
...
     .into_iter()
     .map(|(lt, key)| {
         let mut btn = ctx
             .style()
             .btn_plain
             .icon(lane_type_to_icon(lt).unwrap())
-            .hotkey(if current_lane.is_some() {
-                Some(key.into())
+            .hotkey(if current_lane.is_some() && key.is_some() {
+                Some(key.unwrap().into())
```
[the remark this answers is not in the document]
Comment (dabreegster) [reply game/src/edit/roads.rs:486]: Done!

**Comment by dabreegster** [comment]
> the diagonal striping should go all the way into their containing rect

Screenshot from 2021-07-23 11-19-38
Oh wow this makes such a huge difference in quality!

> the planter's look like their corners are chopped off

Screenshot from 2021-07-23 11-21-44

https://github.com/a-b-street/abstreet/blob/1d7c8ce89cf3180a92da4900bb4083a3d8960089/geom/src/ring.rs#L48

I agree this is annoying. Pre-existing problem also affecting building outlines and anything we hover on. Right now I'm just generating a polyline for the boundary, then doing the normal algorithm to thicken it into a polygon. One hack would be to just stick in an extra polygon to fill in the gap, but it would be a funky `Polygon` whose boundary isn't a nice ring. Proper fix would be to make it a polygon with a hole.. I'll try and improve this separately.

> With this iteration of the jersey barriers, the color does help with the depth perception a bit, but they're still a little hard to distinguish.

These ideas could work. I'll leave this as an open design question for now. If someone makes an example picture, I can make the code to generate it. But I can't even picture what would look good yet.

> that has more to do with the particular photo-realistic asset 

Yeah, the side-view flex post thing was a quick example. Need to find an actual SVG flexpost. I'll look around.

> do all these great new options clutter the UI? Less can be more.

100% agree. I might make the barrier a dropdown, or try to express it in the controls as a specialization of the bike lane.

> The paving on the cycle track should make it clear that the track is off the roadway

There are so many things I love about this image! I'm going to experiment with copying that paving style, the colors, and the kerb style. I think it would improve the sidewalks in A/B Street in general, not just for kerb-protected cycletracks. Will keep you posted, thanks for sharing that Streetcomplete example!

**Comment by dabreegster** [comment]
Thank you everyone for the feedback! Merging this as a first step, will keep exploring rendering style and cleaning up the editor controls.

**Comment by Robinlovelace** [comment]
> 100% agree. I might make the barrier a dropdown, or try to express it in the controls as a specialization of the bike lane.

+1 to that but more importantly amazing work on this epic PR tested, refined, and deployed on the main branch in record time :fire:

**Pull request merged**